jobserver/api.py

```diff
@@ -36,46 +39,67 @@ def post(self, request, *args, **kwargs):
         serializer.is_valid(raise_exception=True)
 
-        job_request_ids = set(JobRequest.objects.values_list("identifier", flat=True))
         incoming_job_request_ids = {j["job_request_id"] for j in serializer.data}
 
+        # error if we find JobRequest IDs in the payload which aren't in the database.
+        job_request_ids = set(JobRequest.objects.values_list("identifier", flat=True))
         missing_ids = incoming_job_request_ids - job_request_ids
         if missing_ids:
             raise ValidationError(f"Unknown JobRequest IDs: {', '.join(missing_ids)}")
 
+        # get JobRequest instances based on the identifiers in the payload
         job_requests = JobRequest.objects.filter(
             identifier__in=incoming_job_request_ids
         )
         job_request_lut = {jr.identifier: jr for jr in job_requests}
 
-        # Remove existing jobs for the JobRequests we've been told about. POSTs
-        # from job-runner are authoritive and declarative so the Jobs in a
-        # given payload are the full set of Jobs the related JobRequests should
-        # know about.
-        Job.objects.filter(job_request__in=job_requests).delete()
-
-        # create Jobs for each object in the payload
-        jobs = []
-        for job in serializer.data:
-            jr_id = job.pop("job_request_id")
-            status = job.pop("status")
-
-            # v1 shim
-            started = job["started_at"] != ""
-            # END
-
-            job_request = job_request_lut[jr_id]
-
-            jobs.append(
-                Job(
-                    job_request=job_request,
-                    runner_status=status,
-                    started=started,
-                    **job,
-                )
-            )
-
-        Job.objects.bulk_create(jobs)
+        # group Jobs by their JobRequest ID
+        jobs_by_request = itertools.groupby(
+            serializer.data, key=operator.itemgetter("job_request_id")
+        )
+        for jr_identifier, jobs in jobs_by_request:
+            # get the JobRequest for this identifier
+            job_request = job_request_lut[jr_identifier]
+
+            # get the current Jobs for the JobRequest, keyed on their identifier
+            jobs_by_identifier = {j.identifier: j for j in job_request.jobs.all()}
+
+            # delete local jobs not in the payload
+            job_request.jobs.exclude(identifier__in=jobs_by_identifier.keys()).delete()
+
+            for job_data in jobs:
+                # this is already set by working through the JobRequest instance
+                job_data.pop("job_request_id")
+
+                # V1 SHIM
+                # we need to translate this runner_status to avoid the .status
+                # property
+                status = job_data.pop("status")
+
+                # we set this based on started_at being set
+                started = bool(job_data["started_at"])
+                # END
+
+                # create new job
```

Comment:
why not `update_or_create()` here?

Reply:
Completely forgot that was a thing, that's signficantly nicer than my version.

Fixed and rebased in 73cf7a4